File: sdks/python/src/attest/engine_manager.py
"""Engine subprocess lifecycle manager."""

from __future__ import annotations

import asyncio
import logging
import os
import shutil
from typing import Any

from attest import __version__
from attest._proto.codec import decode_response, encode_request, extract_result
from attest._proto.types import InitializeParams, InitializeResult

logger = logging.getLogger("attest.engine")

ENGINE_BINARY_NAME = "attest-engine"


def _is_download_disabled() -> bool:
    """Check if auto-download is disabled via ATTEST_ENGINE_NO_DOWNLOAD."""
    val = os.environ.get("ATTEST_ENGINE_NO_DOWNLOAD", "").lower()
    return val in ("1", "true", "yes")


def _find_engine_binary() -> str:
    """Find the engine binary using the full discovery chain.

    Discovery order:
        1. ATTEST_ENGINE_PATH env var (absolute path override)
        2. PATH lookup (shutil.which)
        3. ~/.attest/bin/ shared cache (version-checked)
        4. Monorepo dev layout (../../bin/)
        5. Local ./bin/
        6. Auto-download from GitHub Releases
        7. FileNotFoundError with actionable message
    """
    # Step 1: Explicit env override
    env_path = os.environ.get("ATTEST_ENGINE_PATH")
    if env_path:
        if not os.path.isfile(env_path):
            raise FileNotFoundError(
                f"ATTEST_ENGINE_PATH={env_path} does not exist or is not a file."
            )
        return env_path

    # Step 2: PATH lookup
    found = shutil.which(ENGINE_BINARY_NAME)
    if found:
        return found

    # Step 3: Shared cache (~/.attest/bin/) with version check
    from attest.engine_downloader import cached_engine_path

    cached = cached_engine_path()
    if cached is not None:
        return str(cached)

    # Step 4: Monorepo dev layout
    # Step 5: Local ./bin/
    candidates = [
        os.path.join(os.path.dirname(__file__), "..", "..", "..", "bin", ENGINE_BINARY_NAME),
        os.path.join(os.getcwd(), "bin", ENGINE_BINARY_NAME),
    ]
    for candidate in candidates:
        resolved = os.path.realpath(candidate)
        if os.path.isfile(resolved) and os.access(resolved, os.X_OK):
            return resolved

    # Step 6: Auto-download
    if not _is_download_disabled():
        from attest.engine_downloader import download_engine

        downloaded = download_engine()
        return str(downloaded)

    # Step 7: Actionable error
    raise FileNotFoundError(
        f"Cannot find '{ENGINE_BINARY_NAME}' binary.\n"
        "Install options:\n"
        "  1. Build from source: make engine\n"
        "  2. Download a release: https://github.com/attest-framework/attest/releases\n"
        "  3. Allow auto-download by unsetting ATTEST_ENGINE_NO_DOWNLOAD\n"
        "  4. Set ATTEST_ENGINE_PATH to point to the binary"
    )


class EngineManager:
    """Manages the lifecycle of the attest-engine subprocess."""

    def __init__(
        self,
        engine_path: str | None = None,
        log_level: str = "warn",
    ) -> None:
        self._engine_path = engine_path or _find_engine_binary()
        self._log_level = log_level
        self._process: asyncio.subprocess.Process | None = None
        self._initialized = False
        self._request_id = 0
        self._init_result: InitializeResult | None = None

    async def start(self) -> InitializeResult:
        """Start the engine subprocess and send initialize."""
        self._process = await asyncio.create_subprocess_exec(
            self._engine_path,
            f"--log-level={self._log_level}",
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        logger.info("Engine started (pid=%d)", self._process.pid)

        result = await self._send_request("initialize", InitializeParams(
            sdk_name="attest-python",
            sdk_version=__version__,
            protocol_version=1,
            required_capabilities=["layers_1_4"],
        ).to_dict())

        self._init_result = InitializeResult.from_dict(result)
        if not self._init_result.compatible:
            raise RuntimeError(
                f"Engine incompatible. Missing capabilities: {self._init_result.missing}"
            )
        self._initialized = True
        return self._init_result

    async def stop(self) -> None:
        """Send shutdown and wait for process exit."""
        if self._process is None:
            return
        if self._initialized:
            try:
                await self._send_request("shutdown", {})
            except Exception:
                logger.warning("Shutdown request failed, killing process")
        if self._process.returncode is None:
            self._process.terminate()
            try:
                await asyncio.wait_for(self._process.wait(), timeout=5.0)
            except asyncio.TimeoutError:
                self._process.kill()
                await self._process.wait()
        self._initialized = False
        logger.info("Engine stopped")

    async def send_request(self, method: str, params: dict[str, Any]) -> Any:
        """Send a JSON-RPC request and return the result."""
        if not self._initialized and method != "initialize":
            raise RuntimeError("Engine not initialized. Call start() first.")
        return await self._send_request(method, params)

    async def _send_request(self, method: str, params: dict[str, Any]) -> Any:
        """Internal: send request and read response."""
        assert self._process is not None
        assert self._process.stdin is not None
        assert self._process.stdout is not None

        self._request_id += 1
        request_bytes = encode_request(self._request_id, method, params)

        self._process.stdin.write(request_bytes)
        await self._process.stdin.drain()

        line = await self._process.stdout.readline()
        if not line:
            raise ConnectionError("Engine process closed stdout unexpectedly")

        response = decode_response(line)
        return extract_result(response)

    async def __aenter__(self) -> EngineManager:
        await self.start()
        return self

    async def __aexit__(self, *args: Any) -> None:
        await self.stop()

    @property
    def is_running(self) -> bool:
        return self._process is not None and self._process.returncode is None

File: sdks/typescript/packages/core/src/engine-manager.ts
import { spawn, execSync, type ChildProcess } from "node:child_process";
import { createInterface, type Interface as ReadlineInterface } from "node:readline";
import { existsSync, accessSync, constants as fsConstants } from "node:fs";
import { resolve, join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import type { InitializeResult } from "./proto/types.js";
import { decodeResponse, encodeRequest, extractResult } from "./proto/codec.js";
import { VERSION } from "./version.js";
import { cachedEnginePath, downloadEngine } from "./engine-downloader.js";

const ENGINE_BINARY_NAME = "attest-engine";

async function findEngineBinary(): Promise<string> {
  // 1. Explicit env var override
  const envPath = process.env["ATTEST_ENGINE_PATH"];
  if (envPath) {
    if (!existsSync(envPath)) {
      throw new Error(
        `ATTEST_ENGINE_PATH is set to '${envPath}' but the file does not exist.`
      );
    }
    return envPath;
  }

  // 2. Check PATH via which/where
  try {
    const cmd = process.platform === "win32" ? "where" : "which";
    const found = execSync(`${cmd} ${ENGINE_BINARY_NAME}`, { encoding: "utf-8" }).trim();
    if (found) return found;
  } catch {
    // not on PATH
  }

  // 3. Cached download in ~/.attest/bin/
  const cached = cachedEnginePath();
  if (cached) return cached;

  // 4. Monorepo dev layout (../../bin/ relative to this file)
  // 5. CWD ./bin/
  const candidates = [
    join(dirname(fileURLToPath(import.meta.url)), "..", "..", "..", "..", "..", "bin", ENGINE_BINARY_NAME),
    join(process.cwd(), "bin", ENGINE_BINARY_NAME)
  ];

  for (const candidate of candidates) {
    const resolved = resolve(candidate);
    try {
      accessSync(resolved, fsConstants.X_OK);
      if (existsSync(resolved)) return resolved;
    } catch {
      // not accessible
    }
  }

  // 6. Auto-download
  const noDownload = process.env["ATTEST_ENGINE_NO_DOWNLOAD"];
  if (noDownload === "1" || noDownload === "true" || noDownload === "yes") {
    throw new Error(
      `Cannot find '${ENGINE_BINARY_NAME}' binary and auto-download is disabled ` +
      "(ATTEST_ENGINE_NO_DOWNLOAD is set). Install the engine manually:\n" +
      "  1. Download from https://github.com/attest-framework/attest/releases\n" +
      "  2. Place the binary on your PATH or set ATTEST_ENGINE_PATH"
    );
  }

  return downloadEngine();
}

export class EngineManager {
  private enginePath: string;
  private readonly logLevel: string;
  private process: ChildProcess | null = null;
  private reader: ReadlineInterface | null = null;
  private initialized = false;
  private requestId = 0;
  private initResult: InitializeResult | null = null;

  constructor(options?: { enginePath?: string; logLevel?: string }) {
    this.enginePath = options?.enginePath ?? "";
    this.logLevel = options?.logLevel ?? "warn";
  }

  async start(): Promise<InitializeResult> {
    if (!this.enginePath) {
      this.enginePath = await findEngineBinary();
    }

    this.process = spawn(this.enginePath, [`--log-level=${this.logLevel}`], {
      stdio: ["pipe", "pipe", "pipe"],
    });

    this.reader = createInterface({ input: this.process.stdout! });

    const result = await this.sendRequestInternal("initialize", {
      sdk_name: "attest-typescript",
      sdk_version: VERSION,
      protocol_version: 1,
      required_capabilities: ["layers_1_4"],
      preferred_encoding: "json",
    });

    this.initResult = result as InitializeResult;
    if (!this.initResult.compatible) {
      throw new Error(
        `Engine incompatible. Missing capabilities: ${JSON.stringify(this.initResult.missing)}`,
      );
    }
    this.initialized = true;
    return this.initResult;
  }

  async stop(): Promise<void> {
    if (this.process === null) return;

    if (this.initialized) {
      try {
        await this.sendRequestInternal("shutdown", {});
      } catch {
        // shutdown failed, will kill
      }
    }

    if (this.reader) {
      this.reader.close();
      this.reader = null;
    }

    if (this.process.exitCode === null) {
      this.process.kill("SIGTERM");
      await new Promise<void>((res) => {
        const timer = setTimeout(() => {
          this.process?.kill("SIGKILL");
        }, 5000);
        this.process!.on("exit", () => {
          clearTimeout(timer);
          res();
        });
      });
    }

    this.initialized = false;
    this.process = null;
  }

  async sendRequest(method: string, params: Record<string, unknown>): Promise<unknown> {
    if (!this.initialized && method !== "initialize") {
      throw new Error("Engine not initialized. Call start() first.");
    }
    return this.sendRequestInternal(method, params);
  }

  private async sendRequestInternal(method: string, params: Record<string, unknown>): Promise<unknown> {
    if (this.process === null || this.process.stdin === null || this.reader === null) {
      throw new Error("Engine process not started.");
    }

    this.requestId += 1;
    const requestStr = encodeRequest(this.requestId, method, params);

    return new Promise<unknown>((resolve, reject) => {
      const onLine = (line: string): void => {
        try {
          const response = decodeResponse(line);
          resolve(extractResult(response));
        } catch (err) {
          reject(err);
        }
      };

      this.reader!.once("line", onLine);

      const ok = this.process!.stdin!.write(requestStr, (err: Error | null | undefined) => {
        if (err) {
          this.reader!.removeListener("line", onLine);
          reject(err);
        }
      });

      if (!ok) {
        this.process!.stdin!.once("drain", () => {
          // write will proceed after drain
        });
      }
    });
  }

  get isRunning(): boolean {
    return this.process !== null && this.process.exitCode === null;
  }

  /** Expose process for AttestClient reader loop access */
  get childProcess(): ChildProcess | null {
    return this.process;
  }

  /** Expose readline interface for AttestClient reader loop */
  get readlineInterface(): ReadlineInterface | null {
    return this.reader;
  }
}
